src/Microsoft.Health.Client/Extensions/ServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,90 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Net.Http;
+using EnsureThat;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Client.Configuration;
+using Microsoft.Health.Extensions.DependencyInjection;
+
+namespace Microsoft.Health.Client.Extensions;
+
+public static class ServiceCollectionExtensions
+{
+    public static void AddNamedManagedIdentityCredentialProvider(this IServiceCollection serviceCollection, IConfiguration managedIdentityCredentialConfiguration, string name)
+    {
+        EnsureArg.IsNotNull(serviceCollection, nameof(serviceCollection));
+        EnsureArg.IsNotNull(managedIdentityCredentialConfiguration, nameof(managedIdentityCredentialConfiguration));
+        EnsureArg.IsNotNullOrWhiteSpace(name, nameof(name));
+
+        serviceCollection.Configure<ManagedIdentityCredentialConfiguration>(name, managedIdentityCredentialConfiguration);
+
+        serviceCollection.Add(provider =>
+            {
+                IOptionsMonitor<ManagedIdentityCredentialConfiguration> options = provider.GetService<IOptionsMonitor<ManagedIdentityCredentialConfiguration>>();
+                var credentialProvider = new ManagedIdentityCredentialProvider(options, name);
+                return new NamedCredentialProvider(name, credentialProvider);
+            })
+            .Singleton()
+            .AsSelf();
+    }
+
+    public static void AddNamedOAuth2ClientCertificateCredentialProvider(this IServiceCollection serviceCollection, IConfiguration oAuth2ClientCertificateCredentialConfiguration, string name)
+    {
+        EnsureArg.IsNotNull(serviceCollection, nameof(serviceCollection));
+        EnsureArg.IsNotNull(oAuth2ClientCertificateCredentialConfiguration, nameof(oAuth2ClientCertificateCredentialConfiguration));
+        EnsureArg.IsNotNullOrWhiteSpace(name, nameof(name));
+
+        serviceCollection.Configure<OAuth2ClientCertificateCredentialConfiguration>(name, oAuth2ClientCertificateCredentialConfiguration);
+
+        serviceCollection.Add(provider =>
+            {
+                IOptionsMonitor<OAuth2ClientCertificateCredentialConfiguration> options = provider.GetService<IOptionsMonitor<OAuth2ClientCertificateCredentialConfiguration>>();
+                var httpClient = new HttpClient();
```

Comment:
There seems something weird about creating an `HttpClient` to fetch auth for another `HttpClient`. I am wondering if we should be using something like the `AddHttpMessageHandler` extension method to create our own [`DelegatingHandler`](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.delegatinghandler?view=net-6.0) that wraps these providers to optionally retrieve the token if necessary.

Reply:
Oh! Yes -- of course that's at the end of the registration. And that makes sense about needing to fetch the token.

Reply:
Looking at this service registration, it makes sense that it is singleton, as the token can be used across multiple users' requests. However, I think it'd be ideal for the credential to be singleton (as well as thread-safe!) and long-lived while the actual client is short-lived, as there can be [DNS problems with singleton clients](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests#issues-with-the-original-httpclient-class-available-in-net). Perhaps we use a separate `IHttpClientFactory` for these auth clients? I am unsure if this needs to be addressed immediately, but it's food for thought.